haskell-src/Concordium/Genesis/Parameters.hs

```diff
@@ -266,8 +267,64 @@ instance forall pv. IsProtocolVersion pv => FromJSON (GenesisParameters pv) wher
         unless (any ((facct ==) . gaAddress) gpInitialAccounts) $
             fail $
                 "Foundation account (" ++ show facct ++ ") is not in initialAccounts"
-        return GenesisParameters{..}
+        return GenesisParametersV2{..}
+
+-- | 'GenesisParametersV3' provides a convenient abstraction for
+-- constructing 'GenesisData'. The following invariants are
+-- required to hold:
+--
+-- * There must be at least one baker account in 'gpInitialAccounts'.
+-- * Each baker in 'gpInitialAccounts' must have the correct baker id,
+--   corresponding to its index in the list.
+-- * The foundation account specified in 'gpChainParameters' must
+--   correspond to an account in 'gpInitialAccounts'.
+--
+-- This version is used for consensus version 1.
+data GenesisParametersV3 pv = GenesisParametersV3
+    { -- |Time at which genesis occurs.
+      gp3GenesisTime :: Timestamp,
+      -- |Initial nonce for seeding the leadership election.
+      gp3LeadershipElectionNonce :: LeadershipElectionNonce,
+      -- |Duration of an epoch.
+      gp3EpochDuration :: Duration,
+      -- |Cryptographic parameters.
+      gp3CryptographicParameters :: CryptographicParameters,
+      -- |The identity providers present at genesis.
+      gp3IdentityProviders :: IdentityProviders,
+      -- |The anonymity revokers present at genesis.
+      gp3AnonymityRevokers :: AnonymityRevokers,
+      -- |Initial accounts. Since an account can be a baker, it is important that the
+      -- order of the accounts matches the assigned baker ids.
+      gp3InitialAccounts :: [GenesisAccount],
+      -- |The collection of update keys for performing updates
+      gp3UpdateKeys :: UpdateKeysCollection (AuthorizationsVersionFor (ChainParametersVersionFor pv)),
+      -- |The initial (updatable) chain parameters
+      gp3ChainParameters :: GenesisChainParameters pv
+    }
 
--- |Version number identifying the current version of the genesis parameter format.
-genesisParametersVersion :: Version
-genesisParametersVersion = 2
+instance forall pv. IsProtocolVersion pv => FromJSON (GenesisParametersV3 pv) where
+    parseJSON = withObject "GenesisParametersV3" $ \v -> do
+        gp3GenesisTime <- v .: "genesisTime"
+        gp3LeadershipElectionNonce <- v .: "leadershipElectionNonce"
+        gp3EpochDuration <- v .: "epochDuration"
```

Comment:
Can the epoch duration be 0?

Reply:
No, and the line below disallows this. I have added it as an invariant in the type documentation.